#306 | Ravinder | Adding Accordian Panel for step of type "method".

RavinderSinghMaan opened the pull request:
### Description

Adding Accordion Panel for step of type "method". Request response details are added to Textarea in the Accordion Panel. This will be useful, as it essentially keeps the state of each HTTP call.

- Relevant Issues : #306 
- Type of change :
  - [ ] New feature : Adding Accordian Panel for step of type "method" in UI.

ptrthomas commented on karate-core/src/main/java/com/intuit/karate/ScriptValue.java:
```diff
@@ -265,6 +265,10 @@ public ScriptValue invokeFunction(ScriptContext context) {
+        return getAsPrettyString(true);
+    }
+
+    public String getAsPrettyString(boolean escapeJSON) {
```
I'd like to understand why this change is needed

RavinderSinghMaan replied:
For a request with body like `And request { "id": "http://example.com?id=1", "name": "Billie" }` getAsPrettyString method return escaped JSON and escapes `//` like `\/\/`. New method allows to get just pretty printed JSON not escaped. Please see the screen-shot below how the request looks in log and tooltip in vars window but in raw request response pane it shows just what user passed. 
![image](https://user-images.githubusercontent.com/24655564/36076111-ffcd3802-0f4f-11e8-9ae2-7b9e582e13fa.png)

ptrthomas replied:
ok, it may be a mistake in the way we were doing pretty printing. if you find a better way to escape JSON values (for e.g. if there is a double-quote) what happens, if your approach takes care of this, then we can switch to that for everything

RavinderSinghMaan replied:
Ahh, I can see the actual problem now. Will try to sort it out tomorrow.

RavinderSinghMaan replied:
I have reverted the change in "getAsPrettyString" method. Will raise a separate issue for above problem and will resolve it in separate PR.

ptrthomas commented on karate-core/src/main/java/com/intuit/karate/ScriptValueMap.java:
```diff
@@ -10,17 +10,18 @@
-    public static final String VAR_RESPONSE_HEADERS = "responseHeaders";
-    public static final String VAR_RESPONSE_STATUS = "responseStatus";
-    public static final String VAR_RESPONSE_TIME = "responseTime";
+    public static final String VAR_RESPONSE_COOKIES = VAR_RESPONSE+"Cookies";
```
I know this is debatable, but I'd like to keep this file the way it is

RavinderSinghMaan replied:
Do you mean no new constants in this file or you are not happy with concat?

ptrthomas replied:
not happy with concat. just leave it as it is.

RavinderSinghMaan replied:
done

ptrthomas commented:
Ravinder I got this error:

```
Exception in thread "JavaFX Application Thread" java.lang.NullPointerException
	at com.intuit.karate.ui.StepPanel.lambda$updateRawRequestResponse$5(StepPanel.java:155)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at com.intuit.karate.ui.StepPanel.updateRawRequestResponse(StepPanel.java:155)
	at com.intuit.karate.ui.StepPanel.lambda$run$3(StepPanel.java:145)
	at java.util.Optional.ifPresent(Optional.java:159)
	at com.intuit.karate.ui.StepPanel.run(StepPanel.java:145)
	at com.intuit.karate.ui.StepPanel.action(StepPanel.java:184)
	at com.intuit.karate.ui.ScenarioPanel.action(ScenarioPanel.java:70)
	at com.intuit.karate.ui.SectionPanel.action(SectionPanel.java:89)
	at com.intuit.karate.ui.FeaturePanel.action(FeaturePanel.java:78)
	at com.intuit.karate.ui.AppSession.runAll(AppSession.java:88)
	at com.intuit.karate.ui.HeaderPanel.lambda$new$2(HeaderPanel.java:100)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Node.fireEvent(Node.java:8413)
	at javafx.scene.control.Button.fire(Button.java:185)
	at com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(ButtonBehavior.java:182)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:96)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:89)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:381)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$354(GlassViewEventHandler.java:417)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:416)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.sun.javafx.property.PropertyReference.get(PropertyReference.java:177)
	at javafx.scene.control.cell.PropertyValueFactory.getCellDataReflectively(PropertyValueFactory.java:147)
	at javafx.scene.control.cell.PropertyValueFactory.call(PropertyValueFactory.java:119)
	at javafx.scene.control.cell.PropertyValueFactory.call(PropertyValueFactory.java:98)
	at javafx.scene.control.TableColumn.getCellObservableValue(TableColumn.java:578)
	at javafx.scene.control.TableColumn.getCellObservableValue(TableColumn.java:563)
	at javafx.scene.control.TableCell.updateItem(TableCell.java:644)
	at javafx.scene.control.TableCell.indexChanged(TableCell.java:468)
	at javafx.scene.control.IndexedCell.updateIndex(IndexedCell.java:116)
	at com.sun.javafx.scene.control.skin.TableRowSkinBase.requestCellUpdate(TableRowSkinBase.java:659)
	at com.sun.javafx.scene.control.skin.TableRowSkinBase.lambda$init$497(TableRowSkinBase.java:159)
	at com.sun.javafx.binding.ExpressionHelper$SingleInvalidation.fireValueChangedEvent(ExpressionHelper.java:137)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(ObjectPropertyBase.java:105)
	at javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:112)
	at javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:146)
	at javafx.scene.control.Cell.setItem(Cell.java:403)
	at javafx.scene.control.Cell.updateItem(Cell.java:670)
	at javafx.scene.control.TableRow.updateItem(TableRow.java:259)
	at javafx.scene.control.TableRow.indexChanged(TableRow.java:225)
	at javafx.scene.control.IndexedCell.updateIndex(IndexedCell.java:116)
	at com.sun.javafx.scene.control.skin.VirtualFlow.setCellIndex(VirtualFlow.java:1957)
	at com.sun.javafx.scene.control.skin.VirtualFlow.addTrailingCells(VirtualFlow.java:1344)
	at com.sun.javafx.scene.control.skin.VirtualFlow.layoutChildren(VirtualFlow.java:1197)
	at com.sun.javafx.scene.control.skin.VirtualFlow.setCellCount(VirtualFlow.java:231)
	at com.sun.javafx.scene.control.skin.TableViewSkinBase.updateRowCount(TableViewSkinBase.java:567)
	at com.sun.javafx.scene.control.skin.VirtualContainerBase.checkState(VirtualContainerBase.java:113)
	at com.sun.javafx.scene.control.skin.VirtualContainerBase.layoutChildren(VirtualContainerBase.java:108)
	at com.sun.javafx.scene.control.skin.TableViewSkinBase.layoutChildren(TableViewSkinBase.java:696)
	at javafx.scene.control.Control.layoutChildren(Control.java:578)
	at javafx.scene.Parent.layout(Parent.java:1087)
	at javafx.scene.Parent.layout(Parent.java:1093)
	at javafx.scene.Parent.layout(Parent.java:1093)
	at javafx.scene.Scene.doLayoutPass(Scene.java:552)
	at javafx.scene.Scene$ScenePulseListener.pulse(Scene.java:2397)
	at com.sun.javafx.tk.Toolkit.lambda$runPulse$30(Toolkit.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.Toolkit.runPulse(Toolkit.java:354)
	at com.sun.javafx.tk.Toolkit.firePulse(Toolkit.java:381)
	at com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:510)
	at com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:490)
	at com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$404(QuantumToolkit.java:319)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:71)
	at sun.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:275)
	at com.sun.javafx.property.PropertyReference.get(PropertyReference.java:175)
	... 42 more
Caused by: java.lang.NullPointerException
	at com.intuit.karate.ui.Var.getType(Var.java:51)
	... 52 more
```

steps to reproduce:
1) Start the demo app on port 8080 by running `test.ServerStart` as a JUnit test
2) Run `CatsUiRunner` as a JUnit test, bring up the UI
3) Click the "Run All" button

ptrthomas commented:
I'll also like to propose a change to the UI. Instead of having the accordion panel, how about having a second grid just like the variables but in the position below. would be a nice way to see all HTTP requests. clicking on the method step can optionally highlight the correct row etc. Future feature like generating curl command etc can work off this table. Also filtering, search in future.

<img width="773" alt="screen shot 2018-02-12 at 9 37 21 am" src="https://user-images.githubusercontent.com/915480/36083470-7ebf0004-0fd8-11e8-817d-7519031ff81c.png">

RavinderSinghMaan commented:
@ptrthomas Thanks for taking care of NPE at lower level. I have also refactored the code a bit and added a check for NP. Congrats on new release great work 👍 

IMHO it is really very handy if the request/response details are available next to http calls. The way I was thinking of using UI in our team is purely for API demo at end of sprint. So if raw request/response is available next to the call we won't need to move to different section of the screen. So I was thinking of improving the request/response details in Accordian. I am thinking of adding tabs inside Accordian and will be trying to add json tree. What are your thought on it?

We can also add Network Tab in bottom pane so that user can select Logs or Network?

Also next I will be looking at how to make different panes resizeable? So that user can adjust size of variable or logs pane. If you have any pointer in terms of code that will be useful.

ptrthomas commented:
@RavinderSinghMaan ok the API demo use-case makes sense, please go for it.

Yes something like a network tab would be good in future - that was what I was thinking.

And we totally need panes to be re-sizable, sorry I haven't attempted this yet.
